src/rust/engine/process_execution/src/lib.rs

```diff
@@ -178,59 +183,152 @@ fn serialize_level<S: serde::Serializer>(level: &log::Level, s: S) -> Result<S::
   s.serialize_str(&level.to_string())
 }
 
+/// A symlink from a relative src to an absolute dst (outside of the workdir).
+#[derive(Debug)]
+pub struct WorkdirSymlink {
+  pub src: RelativePath,
+  pub dst: PathBuf,
+}
+
+/// Input Digests for a process execution.
 ///
-/// Input Digests for a process execution. The `complete` Digest is the computed union of all
-/// inputs: the rest of the Digests should be disjoint (or have identical contents where they
-/// overlap).
+/// The `complete` and `nailgun` Digests are the computed union of various inputs.
 ///
+/// TODO: Validate that `immutable_inputs` and `input_files` are disjoint in terms of created
+/// symlinks. Currently the local runner will fail to symlink if the `input_files` contain any
+/// directories/files which collide which the symlinks created for `immutable_inputs`, but we
+/// should fail earlier with an error message. One way to do that would be to create a synthetic
+/// Digest with the symlinks that we will create as files with content like `Symlink to $DIGEST`,
```

Comment:
Side note: REAPI supports symlinks in `Directory` protos: https://github.com/bazelbuild/remote-apis/blob/636121a32fa7b9114311374e4786597d8e7a69f3/build/bazel/remote/execution/v2/remote_execution.proto#L797

Reply:
Interesting, thanks! Definitely didn't think that it would support/encourage absolute links. I think I'll update the comment rather than actually doing this quite yet, but dynamically creating symlink and directory nodes inside of `prepare_workdir` seems like an elegant solution to make both output-directory and symlink creation more rigorous.